enhancements/microshift/periodic-etcd-defragmentation-in-microshift.md

```diff
@@ -0,0 +1,164 @@
+
+### Defragmentation from a goroutine in Microshift
+This solution would launch a goroutine, inside the Microshift binary, that would wait for the condition(s) to run defragmentation.
+This would allow for the logs/status of the defragmentation to appear in the same log stream as the other Microshift logs so and Administrator could see its status and if the defragmentation failed in the same place as they would be monitoring for other Microshift issues/status.
+
+This approach would easily afford the aforementioned periodic/threshold dual condition for defragmentation. It would wait on two timeout channels, one for pure periodic, one (shorter) for calculations; when one fires, defragmentation is started.
+
+#### Defragmentation from cron
+This approach would have the defragmentation condition(s) checked via a cron-job. When Microshift is installed, the cron-job(s) could be added to the tab and the system would handle the timing of the pure periodic and/or the threshold calculation.
+
+There are a few issues with this approach:
+1. It relies on cron being installed on the system and running.
+2. If Microshift is not running, the cronjob might still fire and would fail because it would be able to reach etcd which would cause a lot of false-positive failures to appear in the log.
+3. If we wanted to cover both approaches (pure periodic and threshold), then we would likely need two different cronjobs added to the tab during install.
+
+#### Defragmentation from transient systemd unit
+This would be similar to the work to launch etcd itself as a transient systemd unit. A new sub-command could be added to Microshift (or microshift-etcd) that would do much the same as the goroutine approach above: two timeouts in a loop (assuming doing both pure periodic and threshold) that will launch the defragmentation.
+Its runtime would be attached to that of Microshift and, like the goroutine, it would be easy to detect when/if a defragmentation was running or when/if it failed.
+
+This approach would be the most like an operator in OCP.
+
+### Workflow Description
+For the end user, any change we make should be completely transparent. Defragmentation is only a problem when it isn't set up (or it fails frequently) and etcd runs out of resources, causing a disruption.
+For a Microshift Administrator, they should be able to easily know when a defragmentation is occuring/occured and if it has failed.
+
+#### Variation [optional]
+TODO
+
+### API Extensions
+
+N/A
+
+### Implementation Details/Notes/Constraints [optional]
+
+TODO
+
+### Risks and Mitigations
+
+Some methods - cron-job and systemd unit - could potentially fail silently (or at least force monitoring of a non-Microshift location: cron logs/journalctl) and lead the defragmentation to not execute, which could allow etcd to grow and hit its quota which would cause further writes to fail.
```

Comment:
there's one other mitigation worth mentioning from the CEO auto defrag: kill-switch. I would give the user the ability to opt-out of the defrag altogether, in case it's too disruptive.

Reply:
Good call, I mentioned a vague way of the user being able to configure the defrag timings and thresholds, a good addition to that would be explicitly saying keeping it from running. Thanks

Reply:
I made this explicit in the same section where I discuss the Administrator configuration

Reply:
By "quota" do you mean that etcd has some limits built into itself, or that it would consume too much disk space?

Reply:
That's correct, etcd has a default hard limit on 8GB for the database size. So the size on disk metric measures how much of that quota we've taken up.

Size in use is the active portion of the size on disk; the more different those are from each other, the more fragmented we are.

Defragmentation reduces the size on disk.

Reply:
I wasn't entirely certain about the 8gigs here. This is a setting we supply in CEO: https://github.com/openshift/cluster-etcd-operator/blob/master/pkg/etcdenvvar/etcd_env.go#L37

upstream it seems to be two gigs?
https://github.com/etcd-io/etcd/blob/9bb134233a42c0339cd7c5355dd9c2772e580944/server/storage/quota.go#L27-L34

You should see something like this in the logs: "enabled backend quota with default value"

I didn't find us setting anything else in ushift, so I would actually think it stays at 2gb.

Reply:
Does that size limit need to be user-configurable?

Reply:
Probably wouldn't hurt to allow it to be configurable since etcd won't be able to complete writes if the size limit has been reached.